python/paddle/optimizer/optimizer.py

```diff
@@ -1343,31 +1365,31 @@ def _pir_create_optimization_pass(
 
     def backward(
         self,
-        loss,
-        startup_program=None,
-        parameters=None,
-        no_grad_set=None,
-        callbacks=None,
-    ):
+        loss: Tensor,
+        startup_program: Program | None = None,
+        parameters: list[Tensor] | list[str] | None = None,
+        no_grad_set: set[Tensor] | set[str] | None = None,
+        callbacks: list[Callback] | None = None,
```

Comment:
这里是不是可以展开：

``` python
callbacks: list[Callback[[Block, dict[str, Tensor | paddle.core.OpDesc], None]] | None = None
```

`def append_backward` 里

``` python
        callbacks(list[callable object]|tuple[callable object], optional): List/Tuple of callback functions.
                                               The callbacks are used for
                                               doing some custom jobs during
                                               backward part building. All
                                               callable objects in it will
                                               be invoked once each time a
                                               new gradient operator is added
                                               into the program. The callable
                                               object must have two input
                                               parameters: ``block`` and ``context`` .
                                               The ``block`` is the :ref:`api_guide_Block_en` which
                                               the new gradient operator will
                                               be added to. The ``context`` is a
                                               map, whose keys are gradient
                                               Tensor names and values are
                                               corresponding original :ref:`api_guide_tensor_en` .
                                               In addition to this, the ``context``
                                               has another special key-value pair:
                                               the key is string ``__current_op_desc__``
                                               and the value is the op_desc of the
                                               gradient operator who has just
                                               triggered the callable object.
                                               Default: None.
```

参考 

``` python
def error_clip_callback(block, context):
```

另外，这里只能用 `list`，下面有一句判断 `assert isinstance(callbacks, list)`

@SigureMo 看看是否需要？

Reply:
> callbacks: list[Callback[[Block, dict[str, Tensor | paddle.core.OpDesc], None]] | None = None

Callback 不是泛型，需要改成泛型才能这样写

Reply:
> 另外，这里只能用 list，下面有一句判断 `assert isinstance(callbacks, list)`

这个位置，必须有默认值吧

Reply:
> > callbacks: list[Callback[[Block, dict[str, Tensor | paddle.core.OpDesc], None]] | None = None
> 
> Callback 不是泛型，需要改成泛型才能这样写

改成 Callable 就可以了叭，一开始标错了

Reply:
这里应该是 Callable 还是 Callback 呢？

另外这里有 Block 和 OpDesc，大多是老 IR 的概念了，相关逻辑在 PIR 下是否有适配过呢？

Reply:
> 这里应该是 Callable 还是 Callback 呢？
> 
应该是 Callable，Callback 是模型训练调用的
![image](https://github.com/PaddlePaddle/Paddle/assets/106524776/e3b8c679-06be-4f13-ae42-5665058b3c30)

Reply:
是 Callable ! 

当时在跟踪这个标注的时候，还说要改为 Callable，结果还是忘了 ～ 抱歉 ～～～

> > 另外，这里只能用 list，下面有一句判断 `assert isinstance(callbacks, list)`
> 
> 这个位置，必须有默认值吧

``` python
            if callbacks is None:
                callbacks = [paddle.nn.clip.error_clip_callback]
            else:
                assert isinstance(callbacks, list)

```

代码里面这样处理的，默认是 `None` ，会手动创建 ～